File: src/CmdMark.cpp
// CmdMark.cpp - This file is part of QPlot

/* QPlot - Publication quality 2D graphs with dual coordinate systems
   Copyright (C) 2014  Daniel Wagenaar
  
   This program is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation, either version 3 of the License, or
   (at your option) any later version.
  
   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.
  
   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

// CmdMark.C

#include "CmdMark.h"
#include <math.h>
#include <QPolygonF>
#include "Rotate.h"
#include "Slightly.h"
#include "pi.h"
#include <QDebug>

static CBuilder<CmdMark> cbMark("mark");
static CBuilder<CmdMark> cbPMark("pmark");

bool CmdMark::usage() {
  return error("Usage: mark|pmark xdata ydata [rx [ry [1]]]");
}

static void rendermark(QPainter &p, QPointF const &xy,
		       double r, Marker::Type t, bool asSpine) {
  QPolygonF pf;
  switch (t) {
  case Marker::CIRCLE:
    p.drawEllipse(xy, r, r);
    break;
  case Marker::SQUARE:
    r *= sqrt(PI/4);
    if (asSpine) {
      p.drawLine(QPointF(xy)+QPointF(-r,-r), QPointF(xy)+QPointF(r,r));
      p.drawLine(QPointF(xy)+QPointF(-r,r), QPointF(xy)+QPointF(r,-r));
    } else {
      p.drawRect(QRectF(xy + QPointF(-r,-r), QSizeF(2*r, 2*r)));
    }
    break;
  case Marker::DIAMOND: 
    r *= sqrt(2*PI/4);
    pf.resize(4);
    pf[0] = xy + QPointF(-r, 0);
    pf[1] = xy + QPointF(0, -r);
    pf[2] = xy + QPointF(r, 0);
    pf[3] = xy + QPointF(0, r);
    if (asSpine) {
      p.drawLine(pf[0], pf[2]);
      p.drawLine(pf[1], pf[3]);
    } else {
      p.drawConvexPolygon(pf);
    }
    break;
  case Marker::LEFTTRIANGLE:
    r *= 2./3;
    pf.resize(3);
    pf[0] = xy + QPointF(-2*r, 0);
    pf[1] = xy + QPointF(r, -sqrt(3)*r);
    pf[2] = xy + QPointF(r, sqrt(3)*r);
    if (asSpine) {
      p.drawLine(xy, pf[0]);
      p.drawLine(xy, pf[1]);
      p.drawLine(xy, pf[2]);
    } else {
      p.drawConvexPolygon(pf);
    }
    break;
  case Marker::DOWNTRIANGLE:
    r *= 2./3;
    pf.resize(3);
    pf[0] = xy + QPointF(0, 2*r);
    pf[1] = xy + QPointF(-sqrt(3)*r, -r);
    pf[2] = xy + QPointF(sqrt(3)*r, -r);
    if (asSpine) {
      p.drawLine(xy, pf[0]);
      p.drawLine(xy, pf[1]);
      p.drawLine(xy, pf[2]);
    } else {
      p.drawConvexPolygon(pf);
    }
    break;
  case Marker::RIGHTTRIANGLE:
    r *= 2./3;
    pf.resize(3);
    pf[0] = xy + QPointF(2*r, 0);
    pf[1] = xy + QPointF(-r, sqrt(3)*r);
    pf[2] = xy + QPointF(-r, -sqrt(3)*r);
    if (asSpine) {
      p.drawLine(xy, pf[0]);
      p.drawLine(xy, pf[1]);
      p.drawLine(xy, pf[2]);
    } else {
      p.drawConvexPolygon(pf);
    }
    break;
  case Marker::UPTRIANGLE:
    r *= 2./3;
    pf.resize(3);
    pf[0] = xy + QPointF(0, -2*r);
    pf[1] = xy + QPointF(-sqrt(3)*r, r);
    pf[2] = xy + QPointF(sqrt(3)*r, r);
    if (asSpine) {
      p.drawLine(xy, pf[0]);
      p.drawLine(xy, pf[1]);
      p.drawLine(xy, pf[2]);
    } else {
      p.drawConvexPolygon(pf);
    }
    break;
  case Marker::PENTAGRAM: {
    r *= .5 + .5*sqrt(5);
    double r1 = .5*sqrt(5)-.5;
    r1=r1*r1;
    pf.resize(10);
    for (int k=0; k<10; k++) {
      double dy = -r*cos(2*PI*k/10);
      double dx = r*sin(2*PI*k/10);
      if (k&1) {
	dx *= r1;
	dy *= r1;
      }
      pf[k] = xy + QPointF(dx, dy);
    }
    if (asSpine) {
      for (int k=0; k<10; k+=2)
	p.drawLine(xy, pf[k]);
    } else {
      p.drawPolygon(pf);
    }
  } break;
  case Marker::HEXAGRAM: {
    r=r*sqrt(sqrt(3));
    double r1 = 1/sqrt(3);
    pf.resize(12);
    for (int k=0; k<12; k++) {
      double dy = r*cos(2*PI*k/12);
      double dx = -r*sin(2*PI*k/12);
      if (k&1) {
	dx *= r1;
	dy *= r1;
      }
      pf[k] = xy + QPointF(dx, dy);
    }
    if (asSpine) {
      for (int k=0; k<6; k+=2)
	p.drawLine(pf[k], pf[k+6]);
    } else {
      p.drawPolygon(pf);
    }
  } break;
  case Marker::PLUS:
    p.drawLine(xy + QPointF(r,0), xy + QPointF(-r,0));
    p.drawLine(xy + QPointF(0,r), xy + QPointF(0,-r));
    break;
  case Marker::CROSS:
    r /= sqrt(2);
    p.drawLine(xy + QPointF(r,r), xy + QPointF(-r,-r));
    p.drawLine(xy + QPointF(-r,r), xy + QPointF(r,-r));
    break;
  case Marker::HBAR:
    p.drawLine(xy + QPointF(r,0), xy + QPointF(-r,0));
    break;
  case Marker::VBAR:
    p.drawLine(xy + QPointF(0,r), xy + QPointF(0,-r));
    break;
  case Marker::STAR: {
    double rx = 0.5*r; // cos(60)
    double ry = 0.866*r; // sin(60)
    p.drawLine(xy + QPointF(r,0), xy + QPointF(-r,0));
    p.drawLine(xy + QPointF(rx,ry), xy + QPointF(-rx,-ry));
    p.drawLine(xy + QPointF(rx,-ry), xy + QPointF(-rx,ry));
  } break;
  }
}      
      
bool CmdMark::parse(Statement const &s) {
  int id1 = s.nextIndex(1);
  int id2 = s.nextIndex(id1);
  qDebug() << "parse mark" << s.length() << s.data(1).size() << s.data(id1).size();
  if (!s.isNumeric(1) || !s.isNumeric(id1)
      || s.data(1).size()!=s.data(id1).size())
    return usage();
  if (id2==s.length())
    return true;
  if (s[0].str=="pmark")
    return usage();
  if (!s.isNumeric(id2) || s.data(id2).size()!=1) // rx
    return usage();
  int id3 = s.nextIndex(id2);
  if (id3==s.length())
    return true;
  if (!s.isNumeric(id3) || s.data(id3).size()!=1) // ry
    return usage();
  int id4 = s.nextIndex(id3);
  if (id4==s.length())
    return true;
  if (!s.isNumeric(id4) || s.data(id4).size()!=1) // isvert
    return usage();
  int id5 = s.nextIndex(id4);
  if (id5==s.length())
    return true;
  return usage();
}

QRectF CmdMark::dataRange(Statement const &s) {
  if (s[0].str=="pmark")
    return QRectF(); // paper coord -> no data range

  
  QVector<double> const &xdata = s.data(1);
  QVector<double> const &ydata = s.data(s.nextIndex(1));

  if (xdata.isEmpty())
    return QRectF();

  double minx = xdata[0];
  double maxx = xdata[0];
  double miny = ydata[0];
  double maxy = ydata[0];

  for (int k=1; k<xdata.size(); k++)
    if (xdata[k]<minx)
      minx=xdata[k];
    else if (xdata[k]>maxx)
      maxx=xdata[k];
  for (int k=1; k<ydata.size(); k++)
    if (ydata[k]<miny)
      miny=ydata[k];
    else if (ydata[k]>maxy)
      maxy=ydata[k];
  return QRectF(QPointF(Slightly::less(minx), Slightly::less(miny)),
		QPointF(Slightly::more(maxx), Slightly::more(maxy)));
}

QPolygonF avoidCollision(QPolygonF const &in, double rx, double ry, bool ver) {
  QPolygonF out;
  auto sqr = [](double x) { return x*x; };
  double rx2 = sqr(rx);
  double ry2 = sqr(ry);
  auto collides = [&out, rx2, ry2, sqr](QPointF const &p) {
                    for (QPointF const &q: out) {
                      if (sqr(p.x()-q.x())/rx2 + sqr(p.y()-q.y())/ry2 < 1)
                        return true;
                    }
                    return false;
                  };
  for (QPointF const &p: in) {
    double dx = 0;
    while (collides(p + (ver ? QPointF(0,dx) : QPointF(dx,0))))
      if (dx>0)
        dx = -dx;
      else
        dx = -dx + rx;
    out.append(p + (ver ? QPointF(0,dx) : QPointF(dx,0)));
  }
  return out;
}

    

void CmdMark::render(Statement const &s, Figure &f, bool dryrun) {
  QVector<double> const &xdata = s.data(1);
  int idx = s.nextIndex(1);
  QVector<double> const &ydata = s.data(idx);
  if (xdata.isEmpty()) {
    f.setBBox(QRectF());
    return;
  }
  double rx = 0;
  double ry = 0;
  bool hori = false;
  bool vert = false;
  idx = s.nextIndex(idx);
  if (idx < s.length()) {
    hori = true;
    rx = pt2iu(s.data(idx)[0]);
    ry = rx;
    idx = s.nextIndex(idx);
    if (idx < s.length()) {
      ry = pt2iu(s.data(idx)[0]);
      idx = s.nextIndex(idx);
      if (idx < s.length()) {
        vert = s.data(idx)[0]!=0;
        hori = !vert;
      }
    }
  }
  
  QPolygonF pp(xdata.size());

  if (s[0].str=="mark") {
    for (int k=0; k<xdata.size(); k++)
      pp[k] = f.map(xdata[k], ydata[k]);
    if (hori) 
      pp = avoidCollision(pp, rx, ry, false);
    else if (vert)
      pp = avoidCollision(pp, rx, ry, true);
  } else { // pmark
    double a = f.anchorAngle();
    QPointF xy0 = f.anchor();
    for (int k=0; k<xdata.size(); k++) {
      QPointF xy(pt2iu(xdata[k]), pt2iu(ydata[k]));
      if (a)
	xy = ::rotate(xy, a);
      xy += xy0;
      pp[k] = xy;
    }
  }

  QRectF bbox = pp.boundingRect();
  double w = f.painter().pen().widthF();
  if (w>0)
    bbox.adjust(-w/2, -w/2, w/2, w/2);
  double r = f.marker().radius;
  bbox.adjust(-r, -r, r, r);
  f.setBBox(bbox);
  
  if (dryrun)
    return;

  QPainter &ptr(f.painter());
  ptr.save();

  bool asSpine = false;
  switch (f.marker().fill) {
  case Marker::CLOSED:
    ptr.setBrush(ptr.pen().color());
    break;
  case Marker::OPEN:
    ptr.setBrush(QColor("white"));
    break;
  case Marker::BRUSH:
    break;
  case Marker::SPINE:
    asSpine = true;
    break;
  }

  Marker::Type t = f.marker().type;

  for (QPointF const &p: pp)
    rendermark(ptr, p, r, t, asSpine);

  ptr.restore();  
}

File: backend/python/cmdmark.py
# cmdmark.py - This file is part of QPlot

# QPlot - Publication quality 2D graphs with dual coordinate systems
# Copyright (C) 2014  Daniel Wagenaar
# (same GPL licence as original)

from __future__ import annotations

import math
from typing import TYPE_CHECKING

from PyQt6.QtCore import Qt, QPointF, QRectF, QSizeF
from PyQt6.QtGui import QPainter, QPolygonF, QColor

from .command import Commandfrom .token_ import Tokenfrom .factor import pt2iufrom .rotate import rotate_pointfrom .slightly import slightly_less, slightly_morefrom .marker import Marker
if TYPE_CHECKING:
    from .statement import Statement    from .figure import Figure

# ---------------------------------------------------------------------------
# Module-level helpers
# ---------------------------------------------------------------------------

def _render_mark(p: QPainter, xy: QPointF, r: float,
                 t: Marker.Type, as_spine: bool) -> None:
    """Draw a single marker of type t, centred at xy with radius r."""
    pf: list[QPointF]

    if t == Marker.Type.CIRCLE:
        p.drawEllipse(xy, r, r)

    elif t == Marker.Type.SQUARE:
        r *= math.sqrt(math.pi / 4)
        if as_spine:
            p.drawLine(xy + QPointF(-r, -r), xy + QPointF(r,  r))
            p.drawLine(xy + QPointF(-r,  r), xy + QPointF(r, -r))
        else:
            p.drawRect(QRectF(xy + QPointF(-r, -r), QSizeF(2*r, 2*r)))

    elif t == Marker.Type.DIAMOND:
        r *= math.sqrt(2 * math.pi / 4)
        pf = [
            xy + QPointF(-r,  0),
            xy + QPointF( 0, -r),
            xy + QPointF( r,  0),
            xy + QPointF( 0,  r),
        ]
        if as_spine:
            p.drawLine(pf[0], pf[2])
            p.drawLine(pf[1], pf[3])
        else:
            p.drawConvexPolygon(QPolygonF(pf))

    elif t == Marker.Type.LEFTTRIANGLE:
        r *= 2.0 / 3
        s3r = math.sqrt(3) * r
        pf = [
            xy + QPointF(-2*r,   0),
            xy + QPointF(   r, -s3r),
            xy + QPointF(   r,  s3r),
        ]
        if as_spine:
            for v in pf:
                p.drawLine(xy, v)
        else:
            p.drawConvexPolygon(QPolygonF(pf))

    elif t == Marker.Type.DOWNTRIANGLE:
        r *= 2.0 / 3
        s3r = math.sqrt(3) * r
        pf = [
            xy + QPointF(   0, 2*r),
            xy + QPointF(-s3r,  -r),
            xy + QPointF( s3r,  -r),
        ]
        if as_spine:
            for v in pf:
                p.drawLine(xy, v)
        else:
            p.drawConvexPolygon(QPolygonF(pf))

    elif t == Marker.Type.RIGHTTRIANGLE:
        r *= 2.0 / 3
        s3r = math.sqrt(3) * r
        pf = [
            xy + QPointF( 2*r,   0),
            xy + QPointF(  -r,  s3r),
            xy + QPointF(  -r, -s3r),
        ]
        if as_spine:
            for v in pf:
                p.drawLine(xy, v)
        else:
            p.drawConvexPolygon(QPolygonF(pf))

    elif t == Marker.Type.UPTRIANGLE:
        r *= 2.0 / 3
        s3r = math.sqrt(3) * r
        pf = [
            xy + QPointF(   0, -2*r),
            xy + QPointF(-s3r,    r),
            xy + QPointF( s3r,    r),
        ]
        if as_spine:
            for v in pf:
                p.drawLine(xy, v)
        else:
            p.drawConvexPolygon(QPolygonF(pf))

    elif t == Marker.Type.PENTAGRAM:
        r *= 0.5 + 0.5 * math.sqrt(5)
        r1 = (0.5 * math.sqrt(5) - 0.5) ** 2
        pf = []
        for k in range(10):
            dy = -r * math.cos(2 * math.pi * k / 10)
            dx =  r * math.sin(2 * math.pi * k / 10)
            if k & 1:
                dx *= r1
                dy *= r1
            pf.append(xy + QPointF(dx, dy))
        if as_spine:
            for k in range(0, 10, 2):
                p.drawLine(xy, pf[k])
        else:
            p.drawPolygon(QPolygonF(pf))

    elif t == Marker.Type.HEXAGRAM:
        r = r * math.sqrt(math.sqrt(3))
        r1 = 1.0 / math.sqrt(3)
        pf = []
        for k in range(12):
            dy =  r * math.cos(2 * math.pi * k / 12)
            dx = -r * math.sin(2 * math.pi * k / 12)
            if k & 1:
                dx *= r1
                dy *= r1
            pf.append(xy + QPointF(dx, dy))
        if as_spine:
            for k in range(0, 6, 2):
                p.drawLine(pf[k], pf[k + 6])
        else:
            p.drawPolygon(QPolygonF(pf))

    elif t == Marker.Type.PLUS:
        p.drawLine(xy + QPointF( r, 0), xy + QPointF(-r,  0))
        p.drawLine(xy + QPointF( 0, r), xy + QPointF( 0, -r))

    elif t == Marker.Type.CROSS:
        r /= math.sqrt(2)
        p.drawLine(xy + QPointF( r,  r), xy + QPointF(-r, -r))
        p.drawLine(xy + QPointF(-r,  r), xy + QPointF( r, -r))

    elif t == Marker.Type.HBAR:
        p.drawLine(xy + QPointF(r, 0), xy + QPointF(-r, 0))

    elif t == Marker.Type.VBAR:
        p.drawLine(xy + QPointF(0, r), xy + QPointF(0, -r))

    elif t == Marker.Type.STAR:
        rx = 0.5   * r   # cos(60°)
        ry = 0.866 * r   # sin(60°)
        p.drawLine(xy + QPointF( r,   0), xy + QPointF(-r,   0))
        p.drawLine(xy + QPointF( rx,  ry), xy + QPointF(-rx, -ry))
        p.drawLine(xy + QPointF( rx, -ry), xy + QPointF(-rx,  ry))


def _avoid_collision(pts: list[QPointF], rx: float, ry: float,
                     vertical: bool) -> list[QPointF]:
    """Displace points horizontally (or vertically) to avoid overlap."""
    out: list[QPointF] = []
    rx2 = rx * rx
    ry2 = ry * ry

    def collides(p: QPointF) -> bool:
        for q in out:
            dx = p.x() - q.x()
            dy = p.y() - q.y()
            if dx*dx / rx2 + dy*dy / ry2 < 1:
                return True
        return False

    for p in pts:
        dx = 0.0
        offset = QPointF(0, dx) if vertical else QPointF(dx, 0)
        while collides(p + offset):
            if dx > 0:
                dx = -dx
            else:
                dx = -dx + rx
            offset = QPointF(0, dx) if vertical else QPointF(dx, 0)
        out.append(p + offset)

    return out


# ---------------------------------------------------------------------------
# Command class
# ---------------------------------------------------------------------------

@Command.register("mark")
@Command.register("pmark")
class CmdMark(Command):
    """Draw markers at data or paper coordinates.

    Syntax:
        mark  xdata ydata
        mark  xdata ydata rx [ry [1]]   (displace to avoid crowding)
        pmark xdata ydata
    """

    def _usage(self) -> bool:
        return self.error("Usage: mark|pmark xdata ydata [rx [ry [1]]]")

    def parse(self, s: Statement) -> bool:
        id1 = s.next_index(1)
        id2 = s.next_index(id1)
        if (not s.is_numeric(1) or not s.is_numeric(id1)
                or len(s.data(1)) != len(s.data(id1))):
            return self._usage()
        if id2 == len(s):
            return True
        if s[0].str == "pmark":
            return self._usage()
        if not s.is_numeric(id2) or len(s.data(id2)) != 1:  # rx
            return self._usage()
        id3 = s.next_index(id2)
        if id3 == len(s):
            return True
        if not s.is_numeric(id3) or len(s.data(id3)) != 1:  # ry
            return self._usage()
        id4 = s.next_index(id3)
        if id4 == len(s):
            return True
        if not s.is_numeric(id4) or len(s.data(id4)) != 1:  # isvert
            return self._usage()
        id5 = s.next_index(id4)
        if id5 == len(s):
            return True
        return self._usage()

    def data_range(self, s: Statement) -> QRectF:
        if s[0].str == "pmark":
            return QRectF()  # paper coordinates — no data range

        xdata = s.data(1)
        ydata = s.data(s.next_index(1))
        if not xdata:
            return QRectF()

        return QRectF(
            QPointF(slightly_less(min(xdata)), slightly_less(min(ydata))),
            QPointF(slightly_more(max(xdata)), slightly_more(max(ydata))),
        )

    def render(self, s: Statement, f: Figure, dryrun: bool) -> None:
        xdata = s.data(1)
        idx = s.next_index(1)
        ydata = s.data(idx)

        if not xdata:
            f.set_bbox(QRectF())
            return

        rx = ry = 0.0
        hori = vert = False
        idx = s.next_index(idx)

        if idx < len(s):
            hori = True
            rx = pt2iu(s.data(idx)[0])
            ry = rx
            idx = s.next_index(idx)
            if idx < len(s):
                ry = pt2iu(s.data(idx)[0])
                idx = s.next_index(idx)
                if idx < len(s):
                    vert = s.data(idx)[0] != 0
                    hori = not vert

        if s[0].str == "mark":
            pp = [f.map(xdata[k], ydata[k]) for k in range(len(xdata))]
            if hori:
                pp = _avoid_collision(pp, rx, ry, vertical=False)
            elif vert:
                pp = _avoid_collision(pp, rx, ry, vertical=True)
        else:  # pmark
            a = f.anchor_angle()
            xy0 = f.anchor()
            pp = []
            for k in range(len(xdata)):
                xy = QPointF(pt2iu(xdata[k]), pt2iu(ydata[k]))
                if a:
                    xy = rotate_point(xy, a)
                xy += xy0
                pp.append(xy)

        qpp = QPolygonF(pp)
        bbox = qpp.boundingRect()
        w = f.painter().pen().widthF()
        if w > 0:
            bbox.adjust(-w/2, -w/2, w/2, w/2)
        r = f.marker().radius
        bbox.adjust(-r, -r, r, r)
        f.set_bbox(bbox)

        if dryrun:
            return

        CmdMark.draw(qpp, f)

    @staticmethod
    def draw(pp: QPolygonF, f: Figure) -> None:
        """Draw markers at the given paper-coordinate points.
        Called by CmdHatch as well as by render()."""
        ptr = f.painter()
        ptr.save()

        as_spine = False
        fill = f.marker().fill

        if fill == Marker.Fill.CLOSED:
            ptr.setBrush(ptr.pen().color())
        elif fill == Marker.Fill.OPEN:
            ptr.setBrush(QColor("white"))
        elif fill == Marker.Fill.BRUSH:
            pass  # use current brush as-is
        elif fill == Marker.Fill.SPINE:
            as_spine = True

        t = f.marker().type
        r = f.marker().radius

        if t == Marker.Type.CIRCLE and as_spine:
            ptr.setBrush(Qt.BrushStyle.NoBrush)

        for p in pp:
            _render_mark(ptr, p, r, t, as_spine)

        ptr.restore()
